### PR TITLE
Update pre-commit, store tagger scores as structured array

### LINKED ISSUE #166: Use Ruff and pre-commit.ci
See

- https://beta.ruff.rs/docs/
- https://pre-commit.ci/

### PR DESCRIPTION
**Pull request by samvanstroud**
## Summary

This pull request introduces the following changes

* Update pre-commit consolidating to ruff
* store tagger scores as structured array

Relates to the following issues

* closes #166 

## Conformity
- [x] [Changelog entry](https://github.com/umami-hep/puma/blob/main/changelog.md)
- [x] [Documentation](https://umami-hep.github.io/puma/)

### CODE REVIEW

**Comment by samvanstroud** [comment]
Could I get a review from @afroch @samvanstroud @philippgadow? there are a lot of changes but they are almost all automatic fixes from the updated linting, so I don't think this needs to be scrutinised too closely.

File: docs/sphinx_build_multiversion.py [hunk 1]
```diff
@@ -50,14 +50,12 @@ def build_docs_version(version):
...
-    with open("docs/source/_static/switcher.json", "r") as f:  # pylint: disable=W1514
+    with open("docs/source/_static/switcher.json") as f:
         version_switcher = json.load(f)
 
     # get currently active branch
     command = "git rev-parse --abbrev-ref HEAD".split()
-    initial_branch = (
-        run(command, capture_output=True, check=True).stdout.strip().decode("utf-8")
-    )
+    initial_branch = run(command, capture_output=True, check=True).stdout.strip().decode("utf-8")
```
Comment (afroch) [inline docs/sphinx_build_multiversion.py:58]: Did you change the max length of the lines? It looks like this was changed from 88 to something larger

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,35 +1,24 @@
...
-  - repo: https://github.com/terrencepreilly/darglint
-    rev: v1.8.1
-    hooks:
-      - id: darglint
+  #- repo: https://github.com/pre-commit/mirrors-mypy
+  #  rev: "v1.1.1"
+  #  hooks:
+  #    - id: mypy
+  #      args: [--ignore-missing-imports]
+  #      additional_dependencies: ["types-requests", "types-PyYAML"]
```
Comment (afroch) [inline .pre-commit-config.yaml:18]: We are not using flake8 anymore?

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -1,35 +1,24 @@
...
-        args: [
-            "-rn", # Only display messages
-            "-sn", # Don't display the score
-            "--ignore-paths=.github/*,docs/*",
-          ]
+  #- repo: https://github.com/adrienverge/yamllint.git
+  #  rev: v1.17.0
+  #  hooks:
+  #    - id: yamllint
+  #      args: ["-d", "{extends: relaxed, rules: {line-length: disable}}"]
```
Comment (afroch) [inline .pre-commit-config.yaml:24]: yaml and pylint the same?

**Review by afroch — COMMENTED** [review]
I guess it looks fine to me. Maybe we can set the max line length again to 88. But just a small thing that I wanted to bring up

File: docs/sphinx_build_multiversion.py (hunk 1 above)
Comment (samvanstroud) [reply docs/sphinx_build_multiversion.py:58]: Yeah I changed it to 100, I guess it's kinda arbitrary but I think most people have big enough screens to fit 100 characters, what do you think?

File: .pre-commit-config.yaml (hunk 2 above)
Comment (samvanstroud) [reply .pre-commit-config.yaml:18]: The rules are bundled with ruff :)

File: .pre-commit-config.yaml (hunk 3 above)
Comment (samvanstroud) [reply .pre-commit-config.yaml:24]: yamllint is linting yaml files only. I suggest moving to ruff for all the linting purposes and replace flake8, pylint, etc with it, I've had a good experience doing with with some other packages.

File: docs/sphinx_build_multiversion.py (hunk 1 above)
Comment (afroch) [reply docs/sphinx_build_multiversion.py:58]: On my laptop, when I review the code its a bit too long :smile:
Maybe also to keep it a bit in sync with the rest of the packages, I would opt for 88 again. But not a really strong opinion here. Your call

Comment (samvanstroud) [reply docs/sphinx_build_multiversion.py]: ah ok! I have reverted it to 88 :)

File: puma/histogram.py [hunk 4]
```diff
@@ -476,14 +471,14 @@ def get_discrete_values(self, elem: object):
                 )
         else:
             raise ValueError(
-                "Choose a binning with more than one bin in order to plot"
-                "only discrete values."
+                "Choose a binning with more than one bin in order to plotonly discrete"
```
Comment (dkobylianskii) [inline puma/histogram.py:474]:
```suggestion
                "Choose a binning with more than one bin in order to plot only discrete"
```

**Review by dkobylianskii — COMMENTED** [review]
For me, it looks fine
Thanks, Sam

**Pull request merged**